pkg/upgrade/upgradecluster/tenant_cluster.go

```diff
@@ -120,38 +129,162 @@ import (
 //  5. All tenants have to be at 21.1 cluster version before KV gets upgraded
 //     again in the next release.
 type TenantCluster struct {
-	db *kv.DB
+	// Dialer allows for the construction of connections to other SQL pods.
+	Dialer          NodeDialer
+	InstanceReader  *instancestorage.Reader
+	instancesAtBump []sqlinstance.InstanceInfo
+	DB              *kv.DB
+}
+
+// TenantClusterConfig configures a TenantCluster.
+type TenantClusterConfig struct {
+	// Dialer allows for the construction of connections to other SQL pods.
+	Dialer NodeDialer
+
+	// InstanceReader is used to retrieve all SQL pods for a given tenant.
+	InstanceReader *instancestorage.Reader
+
+	// DB is used to generate transactions for consistent reads of the set of
+	// instances.
+	DB *kv.DB
 }
 
 // NewTenantCluster returns a new TenantCluster.
-func NewTenantCluster(db *kv.DB) *TenantCluster {
-	return &TenantCluster{db: db}
+func NewTenantCluster(cfg TenantClusterConfig) *TenantCluster {
+	return &TenantCluster{
+		Dialer:          cfg.Dialer,
+		InstanceReader:  cfg.InstanceReader,
+		instancesAtBump: make([]sqlinstance.InstanceInfo, 0),
+		DB:              cfg.DB,
+	}
 }
 
-// NumNodes is part of the upgrade.Cluster interface.
-func (t *TenantCluster) NumNodes(ctx context.Context) (int, error) {
-	return 0, errors.AssertionFailedf("non-system tenants cannot iterate nodes")
+// NumNodesOrTenantPods is part of the upgrade.Cluster interface.
+func (t *TenantCluster) NumNodesOrTenantPods(ctx context.Context) (int, error) {
+	// Get the list of all SQL pods running
+	instances, err := t.InstanceReader.GetAllInstances(ctx)
+	if err != nil {
+		return 0, err
+	}
+	return len(instances), nil
 }
 
-// ForEveryNode is part of the upgrade.Cluster interface.
-func (t *TenantCluster) ForEveryNode(
+const BumpClusterVersionOpName = "bump-cluster-version"
+
+// ForEveryNodeOrTenantPod is part of the upgrade.Cluster interface.
+func (t *TenantCluster) ForEveryNodeOrTenantPod(
 	ctx context.Context, op string, fn func(context.Context, serverpb.MigrationClient) error,
 ) error {
-	return errors.AssertionFailedf("non-system tenants cannot iterate nodes")
+	// Get the list of all SQL pods running. Do this using the transactional,
+	// function as transactional protection is required by the tenant upgrade
+	// interlock.
+	instances, err := t.InstanceReader.GetAllInstancesNoCache(ctx)
+	if err != nil {
+		return err
+	}
+
+	// If we're bumping the cluster version, cache the list of instances we
+	// contacted at bump time. This list of instances is then consulted at
+	// fence write time to ensure that we contacted the same set of instances.
+	if strings.Contains(op, BumpClusterVersionOpName) {
+		t.instancesAtBump = instances
+	}
+
+	// Limiting of outgoing RPCs at the tenant level mirrors what we do for
+	// nodes at the storage cluster level.
+	qp := quotapool.NewIntPool("every-sql-pod", 25)
```

Comment:
nit: at least define a constant. In practice it doesn't really cost more to define a tenant read-only cluster setting. I don't anticipate that it'll matter, but I've never seen it be a mistake.